index/internal/indexer/indexer.go

```
@@ -0,0 +1,302 @@
+	// headersMapping will store header paths to a collections.Set of Labels.
+	headersMapping := map[string]*collections.Set[label.Label]{}
+	for _, module := range modules {
+		for _, target := range module.Targets {
+			// Create a targetLabel for the target using the module repository.
+			// It's required to correctly map external module to sources found possibly in other rules
+			targetLabel := label.New(module.Repository, target.Name.Pkg, target.Name.Name)
+			if shouldExcludeTarget(targetLabel) {
+				continue
+			}
+
+			// Normalize headers and add to mapping
+			for hdr := range target.Hdrs {
+				normalizedPath := normalizeHeaderPath(hdr.Name, *target)
+				if shouldExcludeHeader(normalizedPath) {
+					continue
+				}
+				if _, exists := headersMapping[normalizedPath]; !exists {
+					headersMapping[normalizedPath] = &collections.Set[label.Label]{}
+				}
+				headersMapping[normalizedPath].Add(targetLabel)
+			}
+		}
+	}
+
+	// Partition the headers into non-conflicting (exactly one label) and ambiguous (multiple labels).
+	headerToRule := make(map[string]label.Label)
+	ambiguous := make(map[string]collections.Set[label.Label])
+	for path, labels := range headersMapping {
+		switch len(*labels) {
+		case 1:
+			// Extract the only label in the collections.Set.
+			for l := range *labels {
+				headerToRule[path] = l
+				break
+			}
+		default:
+			ambiguous[path] = *labels
+		}
+	}
+
+	return IndexingResult{
+		HeaderToRule: headerToRule,
+		Ambiguous:    ambiguous,
+	}
+}
+
+// Writes the mapping of IndexingResult.HeaderToRule to disk in JSON format.
+// Labels are stored as renered strings
+func (result IndexingResult) WriteToFile(outputFile string) {
+	mappings := make(map[string]string, len(result.HeaderToRule))
+	for hdr, lbl := range result.HeaderToRule {
+		mappings[hdr] = lbl.String()
+	}
+
+	data, err := json.MarshalIndent(mappings, "", "  ")
+	if err != nil {
+		log.Fatalf("Failed to serialize header index to JSON: %v", err.Error())
```

Return error instead of logging.